File: php/chest/ChestFactoryImpl.php
<?php

class Chest {
    private $value;

    public function __construct($value) {
        $this->value = $value;
    }
}

interface ChestFactory {
    public function createBlueChest();
    public function createVioletChest();
    public function createGoldChest();
    public function createRandomChest();
    public function createChestByType($type);
}

class ChestFactoryImpl implements ChestFactory {
    private $random;

    public function __construct() {
        $this->random = mt_rand();
    }

    public function createBlueChest() {
        $low = 1;
        $high = 5;

        return new Chest($low + (int)(($high - $low) * (mt_rand() / mt_getrandmax())));
    }

    public function createVioletChest() {
        $low = 20;
        $high = 50;

        return new Chest($low + (int)(($high - $low) * (mt_rand() / mt_getrandmax())));
    }

    public function createGoldChest() {
        $low = 60;
        $high = 100;

        return new Chest($low + (int)(($high - $low) * (mt_rand() / mt_getrandmax())));
    }

    public function createRandomChest() {
        $k = mt_rand(0, 2);
        if ($k == 0) {
            return $this->createBlueChest();
        } elseif ($k == 1) {
            return $this->createVioletChest();
        } else {
            return $this->createGoldChest();
        }
    }

    public function createChestByType($type) {
        switch ($type) {
            case 'blue':
                return $this->createBlueChest();
            case 'violet':
                return $this->createVioletChest();
            case 'gold':
                return $this->createGoldChest();
            default:
                return $this->createRandomChest();
        }
    }
}
?>

File: java/chest/ChestFactoryImpl.java
package chest;

import java.util.Random;

public class ChestFactoryImpl implements ChestFactory {
    private final Random random;
    public ChestFactoryImpl() {
        random = new Random();
    }

    @Override
    public Chest createBlueChest() {
        int low = 1;
        int high = 5;

        return new Chest(low + (int)((high - low)* random.nextFloat()));
    }

    @Override
    public Chest createVioletChest() {
        int low = 20;
        int high = 50;

        return new Chest(low + (int)((high - low)* random.nextFloat()));
    }

    @Override
    public Chest createGoldChest() {
        int low = 60;
        int high = 100;

        return new Chest(low + (int)((high - low)* random.nextFloat()));
    }

    @Override
    public Chest createRandomChest() {
        Chest chest;
        int k = random.nextInt(3);
        if (k == 0) chest = createBlueChest();
        else if (k == 1) chest = createVioletChest();
        else chest = createGoldChest();

        return chest;
    }

    @Override
    public Chest createChestByType(ChestType type) {
        Chest chest;
        switch (type){
            case blue:
                chest = createBlueChest();
                break;
            case violet:
                chest = createVioletChest();
                break;
            case gold:
                chest = createGoldChest();
                break;
            default:
                chest = createRandomChest();
        }
        return chest;
    }
}
